### PR TITLE
Refactor to support configurations and single entry script

### PR DESCRIPTION
**Pull request by fsschmitt**
- Add configparser support
- Add .ini configuration file
- Add configuration fallback with defaults
- Refactor to support configuration and cli arguments to be expandable 
- Refactor to support group_by functionality based on configuration
- Update test to support single entry script
- Update README.md to support single entry script
- Update README.md to include configuration specs

### CODE REVIEW

File: maas.ini [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# Authentication settings
+[auth]
+# maas_api_key = 
+# maas_api_url = 
+
+# Cache settings
+[cache]
+cache_path = ~/.ansible/tmp
+cache_max_age = 300
```
Comment (mlimaloureiro) [inline maas.ini]: new line

File: maas.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+import uuid
+
+import oauth2 as oauth
+import requests
+import requests_cache
+
+DEFAULTS = {
+    "group_machines_by": "tags",
+    "cache_path": "~/.ansible/tmp",
+    "cache_max_age": 300
```
Comment (mlimaloureiro) [inline maas.py]: cache_max_age_in_{wtv is the unit}

File: maas.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        self.builder = InventoryBuilder()
+
+        # Inventory grouped by hostname or tags
+        self.inventory = self._default_inventory()
+
+        if self.args.list:
+            print(self._list())
+
+    def parse_cli_args(self):
+        ''' Command line argument processing '''
```
Comment (mlimaloureiro) [inline maas.py:194]: what was the reason to remove click ?

File: maas.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+
+class MAASInventory(object):
+
+    def _default_inventory(self) -> dict:
+        return self.builder.build_default()
+
+    def _group_machines_by_tag(self) -> dict:
+        machines = self.fetcher.fetch_machines_grouped_by_tags()
+        return self.builder.build_from_machines(machines)
```
Comment (mlimaloureiro) [inline maas.py]: nl

File: maas.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+    def _group_machines_by_hostname(self) -> dict:
+        machines = self.fetcher.fetch_machines_grouped_by_hostname()
+        return self.builder.build_from_machines(machines)
```
Comment (mlimaloureiro) [inline maas.py]: nl

File: maas.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    def _build_hosts(self, hosts_list: list) -> dict:
+        return {
+            'hosts': hosts_list,
+            'vars': {}
+        }
+
+
+class MAASInventory(object):
+
+    def _default_inventory(self) -> dict:
```
Comment (mlimaloureiro) [inline maas.py]: have the private methods in the bottom

File: maas.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        folder_path = os.path.expanduser(os.path.expandvars(self.cache_path))
+        file_name = os.path.join(folder_path, "ansible-maas-dynamics-inventory")
+
+        if not os.path.exists(folder_path):
+            os.makedirs(folder_path)
+
+        requests_cache.install_cache(file_name, expire_after=self.cache_max_age)
+
+        # Force refresh cache
+        if self.args.refresh_cache and os.path.isfile("{}.sqlite".format(file_name)):
```
Comment (mlimaloureiro) [inline maas.py]: why is sqlite referenced here ? is it caching in sqlite ?

File: maas.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        parser = argparse.ArgumentParser(description='Produce an Ansible Inventory file based on MAAS')
+        parser.add_argument('--list', action='store_true', default=False,
+                            help='List instances (default: True)')
+        parser.add_argument('--host', action='store',
+                            help='Get all the variables about a specific instance')
+        parser.add_argument('--refresh-cache', action='store_true', default=False,
+                            help='Force refresh of cache by making API requests to MAAS (default: False)')
+        self.args = parser.parse_args()
+
+    def read_settings(self):
```
Comment (mlimaloureiro) [inline maas.py]: this method is really big and hard to read

File: maas.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        if not config.has_section('maas'):
+            config.add_section('maas')
+
+        self.group_machines_by = config.get('maas', 'group_machines_by')
+
+    def config_cache(self):
+        ''' Configure the cache settings '''
+
+        folder_path = os.path.expanduser(os.path.expandvars(self.cache_path))
+        file_name = os.path.join(folder_path, "ansible-maas-dynamics-inventory")
```
Comment (mlimaloureiro) [inline maas.py]: constant for `ansible-maas-dynamic-inventory`

File: maas.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            self.inventory = self._group_machines_by_tag()
+        elif self.group_machines_by == "hostnames":
+            self.inventory = self._group_machines_by_hostname()
+
+        return json.dumps(self.inventory, indent=4)
+
+    def __init__(self):
+        """ Main execution path """
+
+        # MAAS Credentials
```
Comment (mlimaloureiro) [inline maas.py]: no need for comments, be explicit on property name instead

File: maas.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        self.credentials = {}
+
+        # Read settings, parse CLI arguments and config Cache
```
Comment (mlimaloureiro) [inline maas.py]: whenever you're adding a comment to a block, it probably means you can find a better name / export to method

File: maas.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        }
+
+    def _build_hosts(self, hosts_list: list) -> dict:
+        return {
+            'hosts': hosts_list,
+            'vars': {}
+        }
+
+
+class MAASInventory(object):
```
Comment (mlimaloureiro) [inline maas.py]: just call it main() it's usually the class that can have a lot of shit there

File: maas.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        self.cache_path = config.get('cache', 'cache_path')
+        self.cache_max_age = config.getint('cache', 'cache_max_age')
+
+        # MAAS Settings
+        if not config.has_section('maas'):
+            config.add_section('maas')
+
+        self.group_machines_by = config.get('maas', 'group_machines_by')
+
+    def config_cache(self):
```
Comment (mlimaloureiro) [inline maas.py]: this doesn't make sense to be here, it's actually in the context of `Fetch` class as a cache provider

**Review by mlimaloureiro — CHANGES_REQUESTED** [review]

File: maas.py (hunk 7 above)
Comment (fsschmitt) [reply maas.py]: Yes, requests_cache by default stores cache in sqlite

File: maas.py (hunk 3 above)
Comment (fsschmitt) [reply maas.py:194]: In order to use the python's standard lib [argparser](https://docs.python.org/3/library/argparse.html)

File: maas.py (hunk 8 above)
Comment (fsschmitt) [reply maas.py]: Broken down into multiple, smaller and more specific methods.

File: maas.py (hunk 13 above)
Comment (fsschmitt) [reply maas.py]: Moved into `Fetch` constructor, configured through the cache_settings variable.

File: maas.py (hunk 9 above)
Comment (fsschmitt) [reply maas.py]: Modified to also be a configurable variable (same as path).

File: maas.py [hunk 14]
```diff
@@ -210,9 +202,15 @@ def parse_cli_args(self):
...
         self.args = parser.parse_args()
 
-    def read_settings(self):
+    def load_settings(self):
         ''' Read configuration settings from ini file and env vars '''
 
+        config = self._load_config_file()
+        self._load_auth_settings(config)
+        self._load_cache_settings(config)
+        self._load_maas_settings(config)
```
Comment (mlimaloureiro) [inline maas.py:211]: so much better 🥇

**Review by mlimaloureiro — APPROVED** [review]

**Pull request merged**